File: 11_SieveOfEratosthenes_CountNonDivisible.cpp
// you can use includes, for example:
// #include <algorithm>

// you can write to stdout for debugging purposes, e.g.
// cout << "this is a debug message" << endl;
#include <algorithm>
#include <cmath>
#include <iostream>
#include <unordered_map>
#include <vector>
vector<int> solution(vector<int> &A) {
  // write your code in C++14 (g++ 6.2.0)
  int n = A.size();

  // 各要素の出現回数を検索
  std::unordered_map<int, int> hash;
  for (int i = 0; i < n; ++i) {
    if (hash.find(A[i]) != hash.end())
      hash[A[i]] += 1;
    else
      hash[A[i]] = 1;
  }

  // // 確認
  // for(auto itr = hash.begin(); itr != hash.end(); itr++){
  //     cout <<itr->first <<" " <<itr->second << endl;
  // }

  // 倍数を足していく
  vector<int> erathos(2 * n + 2, 0);
  int esize = erathos.size();
  for (auto itr = hash.begin(); itr != hash.end(); itr++) {
    int a = itr->first;
    int cnt = itr->second;
    while (a < esize) {
      erathos[a] += cnt;
      a += itr->first;
    }
  }

  // for(int i = 0; i < esize; ++i){
  //     cout <<erathos[i] << " ";
  // }cout << endl;

  vector<int> ans(n);
  for (int i = 0; i < n; ++i) {
    ans[i] = n - erathos[A[i]];
  }

  return ans;
}

File: Python/11_SieveOfEratosthenes_CountNonDivisible.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def solution(A):
    # write your code in Python 3.6
    n = len(A)


    # 各要素の出現回数をカウント
    dic = {}
    for a in A:
        if a in dic: dic[a] += 1
        else: dic[a] = 1

    # エラトステネスの篩みたいな感じで
    # aの倍数にaの出現回数を足してく
    # よって， 0 <= e < (2n + 2) のとき eratos[e] は
    # e % a == 0 となる a∈A の個数を表す
		# エラトステネスの篩の計算量が O(Nloglog(N))なので，
		# それより軽いこのアルゴリズムは大丈夫
    eratos = [0] * (2 *  n + 2) # 1 <= a <= n のため
    for key, value in dic.items():
        k = key
        while k < len(eratos):
            eratos[k] += value    
            k += key
    # end of for

    ans = [0] * n
    for i in range(n):
        ans[i] = n - eratos[A[i]]

    return ans 
# end of func
    


 